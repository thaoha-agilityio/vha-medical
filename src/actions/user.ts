'use server';

import {
  getUsers as getUsersService,
  getUserRoles as getUserRolesService,
  addUser as addUserService,
  updateUser as updateUserService,
} from '@/services';
import { UserPayload } from '@/types';

export const getUsers = async () => {
  const users = await getUsersService();
  return users;
};

export const getUserRoles = async () => {
  const roles = await getUserRolesService();
  return roles;
};

export const addUser = async (data: UserPayload) => {
  const user = await addUserService(data);
  return user;
};

export const updateUser = async (id: string, data: UserPayload) => {
  const user = await updateUserService(id, data);
  return user;
};
