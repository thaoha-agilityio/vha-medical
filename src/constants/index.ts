export * from './apiEndpoint';
export * from './appointment';
export * from './common';
export * from './firebase';
export * from './image';
export * from './message';
export * from './navigation';
export * from './pagination';
export * from './regex';
export * from './router';
export * from './timing';
export * from './chemist';
export * from './encode';
