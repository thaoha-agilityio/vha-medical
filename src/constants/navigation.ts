import {
  AppointmentIcon,
  ChemistIcon,
  DashboardIcon,
  DoctorIcon,
  HospitalIcon,
  MedicineIcon,
  MessageIcon,
  SettingIcon,
} from '@/icons';

// Router
import { PRIVATE_ROUTES } from './router';

export const NAVBAR_LINKS = [
  {
    name: 'Dashboard',
    href: PRIVATE_ROUTES.DASHBOARD,
    icon: DashboardIcon,
    isEnable: true,
  },
  {
    name: 'Appointments',
    href: PRIVATE_ROUTES.APPOINTMENTS,
    icon: AppointmentIcon,
    isEnable: true,
  },
  {
    name: 'Chemists',
    href: PRIVATE_ROUTES.CHEMISTS,
    icon: ChemistIcon,
    isEnable: true,
  },
  { name: 'Hospitals', href: '#', icon: HospitalIcon },
  { name: 'Doctors', href: '#', icon: DoctorIcon },
  { name: 'Medicines', href: '#', icon: MedicineIcon },
];

export const SUPPORT_LINKS = [
  { name: 'Messages', href: '#', icon: MessageIcon },
  {
    name: 'Settings',
    href: '#',
    icon: SettingIcon,
  },
];
