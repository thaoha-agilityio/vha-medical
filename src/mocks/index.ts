export * from './appointments';
export * from './auth';
export * from './chemists';
export * from './common';
export * from './dateTime';
export * from './notification';
export * from './options';
export * from './profile';
export * from './table';
export * from './user';
