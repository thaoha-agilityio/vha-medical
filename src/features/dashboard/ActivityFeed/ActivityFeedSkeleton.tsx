import { memo } from 'react';
import { PAGE_SIZE_DEFAULT } from '@/constants';
import { Card, Skeleton } from '@nextui-org/react';
import { Text } from '@/components/ui';

export const ActivityFeedListSkeleton = memo(() => (
  <div
    className="flex flex-col w-full items-center pt-4"
    data-testid="activity-feed-list-skeleton"
  >
    {Array(PAGE_SIZE_DEFAULT)
      .fill(0)
      .map((_, index) => (
        <div
          key={`activity-feed-skeleton-${index}`}
          className="flex w-full h-[60px] justify-between"
        >
          <Skeleton className="min-w-10 w-10 h-10 rounded-full" />
          <div className="flex flex-col w-full ml-2 mr-8 gap-1">
            <Skeleton className="w-full h-5 rounded-large" />
            <Skeleton className="w-20 h-5 rounded-large" />
          </div>

          <Skeleton className="w-16 h-7 rounded-small mt-3" />
        </div>
      ))}
    <Skeleton className="w-[160px] h-8 rounded-medium mt-6" />
  </div>
));

ActivityFeedListSkeleton.displayName = 'ActivityFeedListSkeleton';

export const ActivityFeedSkeleton = memo(() => (
  <Card className="bg-background-200 p-4 md:pl-7 w-full">
    <Text variant="title" size="lg" customClass="leading-9">
      Activity Feed
    </Text>

    <ActivityFeedListSkeleton />
  </Card>
));

ActivityFeedSkeleton.displayName = 'ActivityFeedSkeleton';
