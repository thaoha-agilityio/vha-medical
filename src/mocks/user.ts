import { AVATAR_THUMBNAIL } from '@/constants';
import {
  ROLE,
  RolePermission,
  UserLogged,
  UserModel,
  UserPayload,
  UserSession,
} from '@/types';

export const MOCK_USER_LIST = {
  data: [
    {
      id: '6',
      attributes: {
        createdAt: '2024-09-20T03:02:27.417Z',
        updatedAt: '2024-09-20T03:02:31.304Z',
        publishedAt: '2024-09-20T03:02:31.303Z',
        users_permissions_user: {
          data: {
            id: '8',
            attributes: {
              username: 'Alex Stanton',
              email: 'alexstanton@gmail.com',
              provider: 'local',
              confirmed: false,
              blocked: false,
              description:
                "Hi, I'm Alex Stanton. Happy for all problem happen with me. Let's smile.",
              rating: 5,
              tasks: 22,
              reviews: 321,
              createdAt: '2024-09-19T02:44:04.643Z',
              updatedAt: '2024-09-19T02:44:04.643Z',
              avatar: {
                data: {
                  id: '10',
                  attributes: {
                    name: 'kyle.webp',
                    alternativeText: null,
                    caption: null,
                    width: 305,
                    height: 305,
                    formats: {
                      thumbnail: {
                        name: 'thumbnail_kyle.webp',
                        hash: 'thumbnail_kyle_554a95ac44',
                        ext: '.webp',
                        mime: 'image/webp',
                        path: null,
                        width: 156,
                        height: 156,
                        size: 7.38,
                        sizeInBytes: 7384,
                        url: '/uploads/thumbnail_kyle_554a95ac44.webp',
                      },
                    },
                    hash: 'kyle_554a95ac44',
                    ext: '.webp',
                    mime: 'image/webp',
                    size: 26.92,
                    url: '/uploads/kyle_554a95ac44.webp',
                    previewUrl: null,
                    provider: 'local',
                    provider_metadata: null,
                    createdAt: '2024-09-19T02:42:30.629Z',
                    updatedAt: '2024-09-24T08:37:10.656Z',
                  },
                },
              },
              role: {
                data: {
                  id: '8',
                  attributes: {
                    name: 'Normal User',
                    description:
                      'Default role given to authenticated user with role normal user.',
                    type: 'normal_user',
                    createdAt: '2024-09-19T02:36:43.012Z',
                    updatedAt: '2024-09-19T02:36:43.012Z',
                  },
                },
              },
            },
          },
        },
      },
    },
    {
      id: 7,
      attributes: {
        createdAt: '2024-09-20T03:02:36.314Z',
        updatedAt: '2024-09-20T03:02:36.898Z',
        publishedAt: '2024-09-20T03:02:36.897Z',
        users_permissions_user: {
          data: {
            id: '9',
            attributes: {
              username: 'Anna White',
              email: 'anna_white@asnet.com.vn',
              provider: 'local',
              confirmed: false,
              blocked: false,
              description: null,
              rating: 5,
              tasks: 40,
              reviews: 600,
              createdAt: '2024-09-19T02:44:45.016Z',
              updatedAt: '2024-09-19T02:44:45.016Z',
              avatar: {
                data: {
                  id: '8',
                  attributes: {
                    name: 'white.webp',
                    alternativeText: null,
                    caption: null,
                    width: 305,
                    height: 305,
                    formats: {
                      thumbnail: {
                        name: 'thumbnail_white.webp',
                        hash: 'thumbnail_white_8b3aabe189',
                        ext: '.webp',
                        mime: 'image/webp',
                        path: null,
                        width: 156,
                        height: 156,
                        size: 7.73,
                        sizeInBytes: 7726,
                        url: '/uploads/thumbnail_white_8b3aabe189.webp',
                      },
                    },
                    hash: 'white_8b3aabe189',
                    ext: '.webp',
                    mime: 'image/webp',
                    size: 28.18,
                    url: '/uploads/white_8b3aabe189.webp',
                    previewUrl: null,
                    provider: 'local',
                    provider_metadata: null,
                    createdAt: '2024-09-19T02:42:30.610Z',
                    updatedAt: '2024-09-24T08:37:10.636Z',
                  },
                },
              },
              role: {
                data: {
                  id: '8',
                  attributes: {
                    name: 'Normal User',
                    description:
                      'Default role given to authenticated user with role normal user.',
                    type: 'normal_user',
                    createdAt: '2024-09-19T02:36:43.012Z',
                    updatedAt: '2024-09-19T02:36:43.012Z',
                  },
                },
              },
            },
          },
        },
      },
    },
    {
      id: '8',
      attributes: {
        createdAt: '2024-09-20T03:02:42.033Z',
        updatedAt: '2024-09-20T03:02:43.413Z',
        publishedAt: '2024-09-20T03:02:43.412Z',
        users_permissions_user: {
          data: {
            id: '7',
            attributes: {
              username: 'Jessica Jane',
              email: 'jessicajane@gmail.com.vn',
              provider: 'local',
              confirmed: false,
              blocked: false,
              description:
                "Hi, I'm Jessica Jane. Lorem Ipsum is simply dummy text of the printing and typesetting for it.",
              rating: 4,
              tasks: 23,
              reviews: 123,
              createdAt: '2024-09-19T02:43:11.055Z',
              updatedAt: '2024-09-19T02:43:11.055Z',
              avatar: {
                data: {
                  id: '7',
                  attributes: {
                    name: 'jane.webp',
                    alternativeText: null,
                    caption: null,
                    width: 305,
                    height: 305,
                    formats: {
                      thumbnail: {
                        name: 'thumbnail_jane.webp',
                        hash: 'thumbnail_jane_bf4e84b9e6',
                        ext: '.webp',
                        mime: 'image/webp',
                        path: null,
                        width: 156,
                        height: 156,
                        size: 7.05,
                        sizeInBytes: 7054,
                        url: '/uploads/thumbnail_jane_bf4e84b9e6.webp',
                      },
                    },
                    hash: 'jane_bf4e84b9e6',
                    ext: '.webp',
                    mime: 'image/webp',
                    size: 26.74,
                    url: '/uploads/jane_bf4e84b9e6.webp',
                    previewUrl: null,
                    provider: 'local',
                    provider_metadata: null,
                    createdAt: '2024-09-19T02:42:30.582Z',
                    updatedAt: '2024-09-24T08:37:10.626Z',
                  },
                },
              },
              role: {
                data: {
                  id: '8',
                  attributes: {
                    name: 'Normal User',
                    description:
                      'Default role given to authenticated user with role normal user.',
                    type: 'normal_user',
                    createdAt: '2024-09-19T02:36:43.012Z',
                    updatedAt: '2024-09-19T02:36:43.012Z',
                  },
                },
              },
            },
          },
        },
      },
    },
    {
      id: '9',
      attributes: {
        createdAt: '2024-09-20T03:02:47.914Z',
        updatedAt: '2024-09-20T03:02:48.891Z',
        publishedAt: '2024-09-20T03:02:48.890Z',
        users_permissions_user: {
          data: {
            id: '10',
            attributes: {
              username: 'John Smith',
              email: 'john_smith@asnet.com.vn',
              provider: 'local',
              confirmed: false,
              blocked: false,
              description: "I'm John Smith from VHA with love.",
              rating: 3,
              tasks: 12,
              reviews: 333,
              createdAt: '2024-09-19T02:46:07.578Z',
              updatedAt: '2024-09-19T02:46:07.578Z',
              avatar: {
                data: {
                  id: '9',
                  attributes: {
                    name: 'alex.webp',
                    alternativeText: null,
                    caption: null,
                    width: 305,
                    height: 305,
                    formats: {
                      thumbnail: {
                        name: 'thumbnail_alex.webp',
                        hash: 'thumbnail_alex_45859439cd',
                        ext: '.webp',
                        mime: 'image/webp',
                        path: null,
                        width: 156,
                        height: 156,
                        size: 7.17,
                        sizeInBytes: 7170,
                        url: '/uploads/thumbnail_alex_45859439cd.webp',
                      },
                    },
                    hash: 'alex_45859439cd',
                    ext: '.webp',
                    mime: 'image/webp',
                    size: 26.72,
                    url: '/uploads/alex_45859439cd.webp',
                    previewUrl: null,
                    provider: 'local',
                    provider_metadata: null,
                    createdAt: '2024-09-19T02:42:30.625Z',
                    updatedAt: '2024-09-24T08:37:10.646Z',
                  },
                },
              },
              role: {
                data: {
                  id: '8',
                  attributes: {
                    name: 'Normal User',
                    description:
                      'Default role given to authenticated user with role normal user.',
                    type: 'normal_user',
                    createdAt: '2024-09-19T02:36:43.012Z',
                    updatedAt: '2024-09-19T02:36:43.012Z',
                  },
                },
              },
            },
          },
        },
      },
    },
    {
      id: '10',
      attributes: {
        createdAt: '2024-09-20T03:03:09.116Z',
        updatedAt: '2024-09-20T03:03:09.580Z',
        publishedAt: '2024-09-20T03:03:09.579Z',
        users_permissions_user: {
          data: {
            id: '6',
            attributes: {
              username: 'ThangHq',
              email: 'thang.hoquang@asnet.com.vn',
              provider: 'local',
              confirmed: false,
              blocked: false,
              description:
                "Hi, I'm Thang Ho Quang. I'm a software developer in AgilityIO. I'm so happy my job.",
              rating: 5,
              tasks: 40,
              reviews: 600,
              createdAt: '2024-09-10T07:07:42.060Z',
              updatedAt: '2024-09-19T02:41:00.646Z',
              avatar: {
                data: {
                  id: '6',
                  attributes: {
                    name: 'avt_hue.jpg',
                    alternativeText: null,
                    caption: null,
                    width: 960,
                    height: 960,
                    formats: {
                      thumbnail: {
                        name: 'thumbnail_avt_hue.jpg',
                        hash: 'thumbnail_avt_hue_335d006883',
                        ext: '.jpg',
                        mime: 'image/jpeg',
                        path: null,
                        width: 156,
                        height: 156,
                        size: 10.36,
                        sizeInBytes: 10359,
                        url: '/uploads/thumbnail_avt_hue_335d006883.jpg',
                      },
                      small: {
                        name: 'small_avt_hue.jpg',
                        hash: 'small_avt_hue_335d006883',
                        ext: '.jpg',
                        mime: 'image/jpeg',
                        path: null,
                        width: 500,
                        height: 500,
                        size: 102.44,
                        sizeInBytes: 102442,
                        url: '/uploads/small_avt_hue_335d006883.jpg',
                      },
                      medium: {
                        name: 'medium_avt_hue.jpg',
                        hash: 'medium_avt_hue_335d006883',
                        ext: '.jpg',
                        mime: 'image/jpeg',
                        path: null,
                        width: 750,
                        height: 750,
                        size: 221.35,
                        sizeInBytes: 221353,
                        url: '/uploads/medium_avt_hue_335d006883.jpg',
                      },
                    },
                    hash: 'avt_hue_335d006883',
                    ext: '.jpg',
                    mime: 'image/jpeg',
                    size: 334.62,
                    url: '/uploads/avt_hue_335d006883.jpg',
                    previewUrl: null,
                    provider: 'local',
                    provider_metadata: null,
                    createdAt: '2024-09-10T07:06:20.012Z',
                    updatedAt: '2024-09-24T08:37:10.613Z',
                  },
                },
              },
              role: {
                data: {
                  id: '6',
                  attributes: {
                    name: 'Admin',
                    description: 'Default role given to authenticated admin.',
                    type: 'authenticated',
                    createdAt: '2024-09-10T03:01:06.879Z',
                    updatedAt: '2024-09-26T09:55:03.464Z',
                  },
                },
              },
            },
          },
        },
      },
    },
  ],
  meta: {
    pagination: {
      page: 1,
      pageSize: 25,
      pageCount: 1,
      total: 5,
    },
  },
};

export const USER_OPTIONS: UserModel[] = [
  {
    id: '1',
    username: 'Alex Stanton',
    email: 'alexstanton@gmail.com',
    description:
      "Hi, I'm Alex Stanton. Happy for all problem happen with me. Let's smile.",
    rating: 5,
    tasks: 22,
    reviews: 321,
    specialtyId: {
      data: {
        id: '1',
        attributes: {
          name: 'Instrumentation',
        },
      },
    },
  },
  {
    id: '2',
    username: 'Anna White',
    email: 'anna_white@asnet.com.vn',
    description: '',
    rating: 5,
    tasks: 40,
    reviews: 600,
    specialtyId: {
      data: {
        id: '2',
        attributes: {
          name: 'Laboratory Chemist',
        },
      },
    },
  },

  {
    id: '3',
    username: 'Jessica Jane',
    email: 'jessicajane@gmail.com.vn',
    description:
      "Hi, I'm Jessica Jane. Lorem Ipsum is simply dummy text of the printing and typesetting for it.",
    rating: 4,
    tasks: 23,
    reviews: 123,
    specialtyId: {
      data: {
        id: '2',
        attributes: {
          name: 'Laboratory Chemist',
        },
      },
    },
  },
];

export const MOCK_USERS_LOGGED: UserLogged[] = [
  {
    id: '30',
    username: 'John Smith',
    email: 'john_smith@asnet.com.vn',
    description: "I'm John Smith from VHA with love.",
    role: { name: ROLE.NORMAL_USER },
    rating: 3,
    tasks: 12,
    reviews: 333,
  },
  {
    id: '20',
    username: 'Jessica Jane',
    email: 'jessica_jane@asnet.com.vn',
    role: { name: ROLE.ADMIN },
    description: "I'm Jessica Jane', from VHA with love.",
    rating: 5,
    tasks: 30,
    reviews: 1442,
  },
];

export const MOCK_USER_ROLE: RolePermission[] = [
  {
    id: 1,
    name: ROLE.NORMAL_USER,
  },
  {
    id: 2,
    name: ROLE.ADMIN,
  },
];

export const MOCK_USER_SESSION: UserSession = {
  id: MOCK_USERS_LOGGED[0].id,
  username: MOCK_USERS_LOGGED[0].username,
  email: MOCK_USERS_LOGGED[0].email,
  token: 'access-token',
  role: '',
  avatar: AVATAR_THUMBNAIL,
  remember: true,
};

export const MOCK_USER_PAYLOAD: UserPayload = {
  ...USER_OPTIONS[0],
  role: 1,
  password: 'mock',
  specialtyId: 1,
};
