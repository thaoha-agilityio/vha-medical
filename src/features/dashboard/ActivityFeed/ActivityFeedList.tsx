'use client';

import { lazy, memo, useTransition } from 'react';
import { Card } from '@nextui-org/react';

// Types
import {
  APIRelatedResponse,
  APIResponse,
  ColumnType,
  InfoModel,
  MetaResponse,
  NotificationModel,
  NotificationResponse,
  UserModel,
} from '@/types';

// Components
import { Avatar, Status, Text } from '@/components/ui';
import {
  formatDateTime,
  fromDateToNow,
  getDescriptionNotification,
} from '@/utils';
import { ActivityFeedListSkeleton } from './ActivityFeedSkeleton';

const DataGrid = lazy(() => import('@/components/ui/DataGrid'));

interface ActivityInfoProps {
  item: NotificationModel;
  userId?: string;
}

const ActivityInfo = memo(({ item, userId = '' }: ActivityInfoProps) => {
  const {
    senderAvatar = '',
    createdAt = '',
    info = {} as InfoModel,
    senderId = {} as APIRelatedResponse<APIResponse<UserModel>>,
  } = item || {};
  const { startTime = '', content } = info;

  const time = formatDateTime(startTime);
  const description = getDescriptionNotification({
    userId,
    senderId,
    time,
    content,
  });

  const timeAgo = fromDateToNow(createdAt);

  return (
    <div className="flex gap-3 justify-items-start ml-1">
      <Avatar isBordered src={senderAvatar} className="aspect-square" />
      <div className="flex flex-col mr-8">
        <Text variant="description" customClass="text-2xs md:text-xs text-wrap">
          {description}
        </Text>
        <Text variant="subTitle" size="2xs">
          {timeAgo}
        </Text>
      </div>
    </div>
  );
});

export const COLUMNS_ACTIVITY_FEED: ColumnType<NotificationModel>[] = [
  {
    key: 'sender',
    title: 'Sender',
  },
  {
    key: 'status',
    title: 'Status',
    customNode: ({ item }) => {
      const { info } = item || {};
      const { status = 0 } = info || {};

      return <Status status={status} className="leading-[27px] max-w-[70px]" />;
    },
  },
];
export interface ActivityFeedListProps extends MetaResponse {
  notifications: NotificationResponse[];
  userId: string;
}

const ActivityFeedList = memo(
  ({ userId, notifications, pagination }: ActivityFeedListProps) => {
    const [isPending, startTransition] = useTransition();

    return (
      <Card as="section" className="bg-background-200 h-fit w-full p-4 md:py-5">
        <Text variant="title" size="lg" customClass="leading-9">
          Activity Feed
        </Text>
        {isPending ? (
          <ActivityFeedListSkeleton />
        ) : (
          <div className="flex flex-col items-center">
            <DataGrid
              id="activity-feed"
              data={notifications}
              pagination={pagination}
              startTransition={startTransition}
              columns={
                COLUMNS_ACTIVITY_FEED.map((column) => ({
                  ...column,
                  customNode: ({ item }) =>
                    column.key === 'sender' ? (
                      <ActivityInfo
                        item={item as NotificationModel}
                        userId={userId}
                      />
                    ) : column.customNode ? (
                      column.customNode({
                        column,
                        item: item as NotificationModel,
                      })
                    ) : null,
                })) as ColumnType<unknown>[]
              }
              classWrapper="pt-4"
              classCell="pb-6"
            />
          </div>
        )}
      </Card>
    );
  },
);

ActivityFeedList.displayName = 'ActivityFeedList';
ActivityInfo.displayName = 'ActivityInfo';
export default ActivityFeedList;
