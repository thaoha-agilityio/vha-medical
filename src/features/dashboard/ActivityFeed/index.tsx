// Types
import { DIRECTION, ROLE, UserLogged } from '@/types';

// Constants
import { API_ENDPOINT, PAGE_SIZE_DEFAULT, PRIVATE_ROUTES } from '@/constants';

// Actions
import { getNotifications } from '@/actions/notification';

// Components
import ActivityFeedList from './ActivityFeedList';

export interface ActivityFeedProps {
  page: number;
  userLogged: UserLogged | null;
}

const ActivityFeed = async ({ page, userLogged }: ActivityFeedProps) => {
  const { id: userId = '', role: roleModel } = userLogged || {};
  const { name: role = ROLE.NORMAL_USER } = roleModel || {};

  const searchParamsAPI = () => {
    const params = new URLSearchParams();
    params.set('populate[0]', 'senderId');
    params.set('pagination[page]', page.toString());
    params.set('pagination[pageSize]', PAGE_SIZE_DEFAULT.toString());
    params.set('sort[0]', `createdAt:${DIRECTION.DESC}`);

    if (role === ROLE.NORMAL_USER || !role) {
      params.set('filters[senderId][id][$eq]', `${userId}`);
    }
    return params;
  };

  const { notifications, error, ...meta } = await getNotifications({
    searchParams: searchParamsAPI(),
    options: {
      next: {
        tags: [
          API_ENDPOINT.NOTIFICATIONS,
          `${PRIVATE_ROUTES.DASHBOARD}/${userId}`,
        ],
      },
    },
  });

  if (error) throw error;

  return (
    <ActivityFeedList
      userId={userId}
      notifications={notifications || []}
      pagination={meta.pagination && { ...meta.pagination, page: Number(page) }}
    />
  );
};

ActivityFeed.displayName = 'ActivityFeed';
export default ActivityFeed;
