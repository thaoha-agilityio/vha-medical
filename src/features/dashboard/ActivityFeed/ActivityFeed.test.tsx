import React, { act } from 'react';
import { usePathname, useRouter, useSearchParams } from 'next/navigation';
import ActivityFeed, { ActivityFeedProps } from '.';
import { getNotifications } from '@/services';
import { MOCK_NOTIFICATION_LIST } from '@/mocks';
import { PRIVATE_ROUTES } from '@/constants';
import { MOCK_USERS_LOGGED } from '@/mocks/user';
import { renderServerComponent } from '@/utils/test-util';

jest.mock('@/services/notification.ts', () => ({
  getNotifications: jest.fn(),
}));

jest.mock('next/navigation', () => ({
  usePathname: jest.fn(),
  useSearchParams: jest.fn(),
  useRouter: jest.fn(),
}));

describe('ActivityFeed test cases', () => {
  const mockProps: ActivityFeedProps = {
    page: 1,
    userLogged: MOCK_USERS_LOGGED[0],
  };
  const mockReplace = jest.fn();
  const mockGetNotifications = getNotifications as jest.Mock;
  const mockUsePathname = usePathname as jest.Mock;
  const mockUseRouter = useRouter as jest.Mock;
  const mockUseSearchParams = useSearchParams as jest.Mock;

  const setup = async (props: ActivityFeedProps) =>
    act(() => renderServerComponent(<ActivityFeed {...props} />));

  beforeEach(() => {
    jest.spyOn(React, 'useMemo').mockReturnValueOnce({});
    mockUsePathname.mockReturnValue(PRIVATE_ROUTES.DASHBOARD);
    mockUseRouter.mockReturnValue({ replace: mockReplace });
    mockUseSearchParams.mockReturnValue(new URLSearchParams());
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly with props', async () => {
    mockGetNotifications.mockResolvedValueOnce(MOCK_NOTIFICATION_LIST);

    const { asFragment } = await setup(mockProps);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly without empty role', async () => {
    mockGetNotifications.mockResolvedValueOnce(MOCK_NOTIFICATION_LIST);

    const { asFragment } = await setup({ ...mockProps, userLogged: null });

    expect(asFragment()).toMatchSnapshot();
  });
});
