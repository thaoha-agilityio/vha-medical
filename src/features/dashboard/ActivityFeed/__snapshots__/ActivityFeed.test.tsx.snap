// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityFeed test cases should render correctly with props 1`] = `
<DocumentFragment>
  <section
    class="flex flex-col relative overflow-hidden text-foreground box-border outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none bg-background-200 h-fit w-full p-4 md:py-5"
    tabindex="-1"
  >
    <p
      class="text-primary-100 font-semibold text-lg leading-9"
    >
      Activity Feed
    </p>
    <div
      class="flex flex-col items-center"
    >
      <div
        class="flex flex-col relative gap-4 w-full"
      >
        <div
          class="z-0 flex flex-col relative justify-between gap-4 overflow-auto rounded-large w-full bg-transparent-200 shadow-none p-0 pt-4"
        >
          <table
            aria-describedby=""
            aria-label="Table"
            class="min-w-full h-auto table-auto w-full"
            id="activity-feed"
            role="grid"
            tabindex="0"
          >
            <thead
              class="[&>tr]:first:rounded-lg hidden"
              role="rowgroup"
            >
              <tr
                class="group outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2"
                role="row"
              >
                <th
                  class="group px-3 h-10 align-middle bg-default-100 whitespace-nowrap text-foreground-500 text-tiny font-semibold first:rounded-s-lg last:rounded-e-lg data-[sortable=true]:cursor-pointer data-[hover=true]:text-foreground-400 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-start"
                  data-key="sender"
                  id="activity-feed-sender"
                  role="columnheader"
                  tabindex="-1"
                >
                  Sender
                </th>
                <th
                  class="group px-3 h-10 align-middle bg-default-100 whitespace-nowrap text-foreground-500 text-tiny font-semibold first:rounded-s-lg last:rounded-e-lg data-[sortable=true]:cursor-pointer data-[hover=true]:text-foreground-400 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-start"
                  data-key="status"
                  id="activity-feed-status"
                  role="columnheader"
                  tabindex="-1"
                >
                  Status
                </th>
              </tr>
              <tr
                aria-hidden="true"
                class="w-px h-px block"
                style="margin-left: 0.25rem; margin-top: 0.25rem;"
                tabindex="-1"
              />
            </thead>
            <tbody
              data-empty="true"
              role="rowgroup"
            >
              <tr
                role="row"
              >
                <td
                  class="align-middle text-center text-primary-100 text-xl font-medium h-[152px] pb-4"
                  colspan="2"
                  role="gridcell"
                >
                  Result Not Found
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`ActivityFeed test cases should render correctly without empty role 1`] = `
<DocumentFragment>
  <section
    class="flex flex-col relative overflow-hidden text-foreground box-border outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none bg-background-200 h-fit w-full p-4 md:py-5"
    tabindex="-1"
  >
    <p
      class="text-primary-100 font-semibold text-lg leading-9"
    >
      Activity Feed
    </p>
    <div
      class="flex flex-col items-center"
    >
      <div
        class="flex flex-col relative gap-4 w-full"
      >
        <div
          class="z-0 flex flex-col relative justify-between gap-4 overflow-auto rounded-large w-full bg-transparent-200 shadow-none p-0 pt-4"
        >
          <table
            aria-describedby=""
            aria-label="Table"
            class="min-w-full h-auto table-auto w-full"
            id="activity-feed"
            role="grid"
            tabindex="0"
          >
            <thead
              class="[&>tr]:first:rounded-lg hidden"
              role="rowgroup"
            >
              <tr
                class="group outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2"
                role="row"
              >
                <th
                  class="group px-3 h-10 align-middle bg-default-100 whitespace-nowrap text-foreground-500 text-tiny font-semibold first:rounded-s-lg last:rounded-e-lg data-[sortable=true]:cursor-pointer data-[hover=true]:text-foreground-400 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-start"
                  data-key="sender"
                  id="activity-feed-sender"
                  role="columnheader"
                  tabindex="-1"
                >
                  Sender
                </th>
                <th
                  class="group px-3 h-10 align-middle bg-default-100 whitespace-nowrap text-foreground-500 text-tiny font-semibold first:rounded-s-lg last:rounded-e-lg data-[sortable=true]:cursor-pointer data-[hover=true]:text-foreground-400 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-start"
                  data-key="status"
                  id="activity-feed-status"
                  role="columnheader"
                  tabindex="-1"
                >
                  Status
                </th>
              </tr>
              <tr
                aria-hidden="true"
                class="w-px h-px block"
                style="margin-left: 0.25rem; margin-top: 0.25rem;"
                tabindex="-1"
              />
            </thead>
            <tbody
              data-empty="true"
              role="rowgroup"
            >
              <tr
                role="row"
              >
                <td
                  class="align-middle text-center text-primary-100 text-xl font-medium h-[152px] pb-4"
                  colspan="2"
                  role="gridcell"
                >
                  Result Not Found
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
